### ISSUE TITLE
CherryPy splits query arguments on semicolon

### ISSUE THREAD

**Issue by datendelphin** [cherrypy/cherrypy#1860]
**I'm submitting a ...**
- [ ] bug report
- [ ] feature request
- [ ] question about the decisions made in the repository

**What is the current behavior?**
Given the following query /?example=0;1;3&another=2;4
kwargs reads {'example': '0', '1': '', '3': '', 'another': '2', '4': ''}

**What is the expected behavior?**
kwargs should be {'example': '0;1;3', 'another': '2;4'}

**Please tell us about your environment:**

- CherryPy version: 18.6.0
- Python version: 3.8.3
- OS: Linux

**Other information**
Seems related, but hard to follow: #1778

**Comment by webknjaz** [comment]
So the problem is that older W3C recommendation related to HTML 4.01 recommended supporting a semicolon as an alias to an ampersand in 1999:

_https://www.w3.org/TR/1999/REC-html401-19991224/appendix/notes.html#h-B.2.2_:
> We recommend that HTTP server implementors, and in particular, CGI implementors support the use of ";" in place of "&" to save authors the trouble of escaping "&" characters in this manner.

There's a newer recommendation about HTML 5 which does not mention the semicolon anymore:
> To decode application/x-www-form-urlencoded payloads, the following algorithm should be used. [...] The output of this algorithm is a sorted list of name-value pairs. [...]
> 1. Let strings be the result of strictly splitting the string payload on U+0026 AMPERSAND characters (&).

OTOH these documents mention how to process and send data using forms. In particular, using `application/x-www-form-urlencoded` encoding method (there's also multipart!). And when using forms, their data may be passed either via GET params or via POST. So there's also a case where GET params in URLs are just that, not something that got there from forms. In this case, why would those documents apply here?

Also, [some people on SO seem to think that the recommendation from 1999 is obsolete](https://stackoverflow.com/a/40768572/595220). But then, it's a recommendation about HTML, not HTTP. The web page authors can use one or another. And in case they rely on HTML 4.01, the web-server should support it, right?

FWIW the implementation is here: https://github.com/cherrypy/cherrypy/blob/b57bdc5/cherrypy/lib/httputil.py#L344. But I'm not sure if it deserves to be changed.

By the way, have you tried escaping your semicolons as `%3B` (which is a urlencoded value for `;`)?

**Comment by webknjaz** [comment]
FTR the built-in python's `urllib.parse.parse_qs()` even drops things starting `;` and before the following `&` for good:
```python
$ ipython
Python 3.7.1 (default, Jan 28 2019, 08:25:13) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.1.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import urllib                                                         

In [2]: urllib.parse.parse_qs(urllib.parse.urlsplit('/?example=0;1;3&another=2;4').query)                                                           
Out[2]: {'example': ['0'], 'another': ['2']}

In [3]: urllib.parse.parse_qs(urllib.parse.urlsplit('/?example=0%3B1%3B3&another=2%3B4').query)                                                     
Out[3]: {'example': ['0;1;3'], 'another': ['2;4']}
```

**Comment by datendelphin** [comment]
Thank you @webknjaz for this insightful explanation. I was unaware that ";" was defined to have the same meaning as "&". I agree that this behaviour makes sense and this is not a bug.

As for the correct solution to my problem (to interpret the query string for an api that uses ";" in the values) I found that using cherrypy.request.query_string is the better approach for this quirky api.

**Comment by webknjaz** [comment]
Could you please clarify FTR: have urlencoding the semicolon worked for you?

**Comment by datendelphin** [comment]
Indeed it does work. But due to the nature of the application, I have unfortunately no control over the side generating the query.

**Comment by webknjaz** [comment]
Oh... too bad. Monkey-patching things is not a public API, hence not guaranteed to work across versions.

I think you could use a [`on_start_resource`](https://docs.cherrypy.org/en/latest/extend.html#hookpoint) [tool](https://docs.cherrypy.org/en/latest/extend.html#tools) that rewrites a query string for the current request:

```python
#! /usr/bin/env python3

import cherrypy


@cherrypy.tools.register('on_start_resource')
def amp_only_query_string():
    patched_qs = cherrypy.request.query_string.replace(';', '%3B')
    cherrypy.request.query_string = patched_qs


class HelloWorld:
    @cherrypy.expose
    @cherrypy.tools.amp_only_query_string()
    def index(self, **params):
        return (
            f'Handler params: {params!s}<br>\n'
            f'Query string: {cherrypy.request.query_string}'
        )


if __name__ == '__main__':
    cherrypy.quickstart(HelloWorld(), '/')
```

You can also achieve this on the dispatcher level, I think: https://docs.cherrypy.org/en/latest/extend.html#tailored-dispatchers.

**Comment by webknjaz** [comment]
For the sake of documenting the HTTP aspect of this (for history), here's a few quotes from RFCs. Let's start with [RFC 7230](https://tools.ietf.org/html/rfc7230):

[HTTP request starts with a `request-line`](https://tools.ietf.org/html/rfc7230#section-3.1.1) which is defined as
>      request-line   = method SP request-target SP HTTP-version CRLF

We're interested in the [`request-target`](https://tools.ietf.org/html/rfc7230#section-5.3). So let's look it up:
>     request-target = origin-form
>                    / absolute-form
>                    / authority-form
>                    / asterisk-form

The option we're interested in is [`origin-form`](https://tools.ietf.org/html/rfc7230#section-5.3.1) which is defined as
>      origin-form    = absolute-path [ "?" query ]

[RFC 7230, appendix B](https://tools.ietf.org/html/rfc7230#appendix-B) sends has a summary of definitions:
>   query = <query, see [RFC3986], Section 3.4>

So it basically sends us over to [RFC 3986, section 3.4](https://tools.ietf.org/html/rfc3986#section-3.4) which says
>       query       = *( pchar / "/" / "?" )

What's left is to look up [`pchar` in section 3.3](https://tools.ietf.org/html/rfc3986#section-3.3):
>       pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"

Almost there! I'm going to paste less important parts 
>      unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
>      pct-encoded = "%" HEXDIG HEXDIG

(where HEXDIG is a hexadecimal digit)

The interesting part is [`sub-delims`, which is defined in section 2.2](https://tools.ietf.org/html/rfc3986#section-2.2):
>      reserved    = gen-delims / sub-delims
>
>      gen-delims  = ":" / "/" / "?" / "#" / "[" / "]" / "@"
>
>      sub-delims  = "!" / "$" / "&" / "'" / "(" / ")"
>                  / "*" / "+" / "," / ";" / "="

As you see, `;` is listed alongside `&`, in the same category.

What's even more interested is the explanation in this section:
>    URIs include components and subcomponents that are delimited by
>   characters in the "reserved" set.  These characters are called
>   "reserved" because **they may (or may not) be defined as delimiters** by
>   the generic syntax, by each scheme-specific syntax, or by the
>   implementation-specific syntax of a URI's dereferencing algorithm.
>   **If data for a URI component would conflict with a reserved
>   character's purpose as a delimiter, then the conflicting data must be
>   percent-encoded before the URI is formed.**

This basically says that it's the client's responsibility to escape ambiguous chars.

Going further:

>   The purpose of reserved characters is to provide a set of delimiting
>   characters that are distinguishable from other data within a URI.
>   URIs that differ in the replacement of a reserved character with its
>   corresponding percent-encoded octet are not equivalent.  **Percent-
>   encoding a reserved character, or decoding a percent-encoded octet
>   that corresponds to a reserved character, will change how the URI is
>   interpreted by most applications.**  Thus, **characters in the reserved
>   set are protected from normalization and are therefore safe to be
>   used by scheme-specific and producer-specific algorithms for
>   delimiting data subcomponents** within a URI.
>
>   [..]
>
>   **URI producing applications should percent-encode data octets that
>   correspond to characters in the reserved set unless these characters
>   are specifically allowed by the URI scheme to represent data in that
>   component.**  If a reserved character is found in a URI component and
>   no delimiting role is known for that character, then it must be
>   interpreted as representing the data octet corresponding to that
>   character's encoding in US-ASCII.

The last sentence basically says that if you decide that in your app a reserved char is not a delimiter, it's okay to adjust your parser accordingly.

**Comment by mcow** [comment]
@webknjaz Thank you for tagging me on this issue. I haven't looked at this since I closed out that other bug, but I spun everything up again. To my surprise, whatever the symptom I was complaining about then, I see no problem with CherryPy now. I do see an error within jQuery that happens _after_ I've processed the query, but the fields of the query are properly parsed into a dict and handed to my dispatch function exactly as I expected. I can only assume I mis-read the symptoms at the time. Sorry for the bother.
